packages/web-components/src/tree-item/tree-item.ts

```diff
@@ -0,0 +1,167 @@
+   * @public
+   * @remarks
+   * HTML Attribute: selected
+   */
+  @attr({ mode: 'boolean' })
+  selected = false;
+  protected selectedChanged(): void {
+    if (this.$fastController.isConnected) {
+      this.$emit('selected-change', this);
+    }
+  }
+
+  /**
+   * The size of the tree item element
+   */
+  @observable
+  size: 'medium' | 'small' = 'medium';
+  private sizeChanged() {
+    this.updateChildTreeItems();
+  }
+
+  /**
+   * The size of the tree item element
+   */
+  @observable
+  appearance: 'subtle' | 'subtle-alpha' | 'transparent' = 'subtle';
+  private appearanceChanged() {
+    this.updateChildTreeItems();
+  }
+
+  @observable
+  childTreeItems: TreeItem[] = [];
+  private childTreeItemsChanged() {
+    this.updateChildTreeItems();
+  }
+
+  /**
+   * 1. Update the child items' size based on the tree-view's size
+   * 2. Update the child items' appearance based on the tree-view's appearance
+   */
+  private updateChildTreeItems() {
+    if (!this.childTreeItems || !this.childTreeItems.length) {
+      return;
+    }
+    this.childTreeItems.forEach(item => {
+      item.size = this.size;
+      item.appearance = this.appearance;
+    });
+  }
+
+  /**
+   * Handle focus events
+   *
+   * @internal
+   */
+  handleFocus = (e: FocusEvent): void => {
+    if (
+      e.target === this ||
+      // In case where the tree-item contains a focusable element, we should not set the tabindex to 0 when the focus is on its child focusable element,
+      // so users can shift+tab to navigate to the tree-item from its child focusable element.
+      this.contains(e.target as Node)
+    ) {
+      this.setAttribute('tabindex', '0');
+    }
+  };
+
+  /**
+   * Handle blur events
+   *
+   * @internal
+   */
+  handleBlur = (e: FocusEvent): void => {
+    if (e.target === this) {
+      this.setAttribute('tabindex', '-1');
+    }
+  };
+
+  /**
+   * Toggle the expansion state of the tree item
+   */
+  toggleExpansion() {
+    if (!this.disabled && this.childTreeItems.length) {
+      this.expanded = !this.expanded;
+    }
+  }
+
+  /**
+   * Toggle the single selection state of the tree item
+   */
+  toggleSelection() {
+    if (!this.disabled) {
+      this.selected = !this.selected;
+    }
+  }
+
+  /**
+   * Whether the tree is nested
+   * @internal
+   */
+  get isNestedItem() {
+    return this.parentElement instanceof TreeItem;
+  }
+
+  /**
+   * Whether the tree is an root item
+   * @internal
+   */
+  get isRootItem() {
+    return this.parentElement && (this.parentElement as any).isTreeView;
+  }
+
+  /**
+   * Calculate the class list that should be applied to the positioning-region
+   * @internal
+   */
+  @volatile
+  get calculatedClassName() {
+    let className = '';
+    if (!this.childTreeItems.length) {
+      className += 'leaf';
+    }
+    if (this.isRootItem) {
+      className += ' root-item';
+    }
+    className += ` ${this.size} ${this.appearance}`;
```

Comment:
If seems like these could be handled with the parent attribute selector, though I might be missing something. Do we need to modify classes if we can use the cascade and specificity?

Reply:
`isRootItem` is from parent attribute, which is handled from outer elements.
The other classNames is necessary to be assembled here.